### PR TITLE
#30275 Task Upgrade to create and popualte new unique_fields table

### LINKED ISSUE #30275: Create Task Upgrade to create new unique_field table
### Parent Issue

#29459

### User Story

According to spike #29707, the best solution to address the unique fields bug is to create a new table to store and validate all unique values. To implement this, we need to create a new Task Upgrade to create and populate this new unique_fields table.

**New Table Structure**

- unique_key_val (bigint): A hash calculated by concatenating the following values: content_type, field_var, host_id, language, and value. The host_id is only included if the field variable uniquePerSite is set.
- supporting_values (jsonb): A JSON object containing the following fields: content_type, field_var, host_id, language, value, uniquePerSite, and content_identifier. The content_identifier will be an array, as multiple contentlets may share the same value before the upgrade.

The unique_key_val is the PK of the table.

**Populate New Table**

We need to populate this new table with the unique field values that already exist in the system. The following query was used during the spike to retrieve these unique values:
```

SELECT DISTINCT 
    structure.velocity_var_name AS type_var_name,
    field.velocity_var_name AS field_var_name,
    contentlet.language_id,
    identifier.host_inode,
    jsonb_extract_path_text(contentlet_as_json->'fields', field.velocity_var_name)::jsonb->>'value' AS field_value
FROM contentlet
    INNER JOIN structure ON structure.inode = contentlet.structure_inode
    INNER JOIN field ON structure.inode = field.structure_inode
    INNER JOIN identifier ON contentlet.identifier = identifier.id
WHERE jsonb_extract_path_text(contentlet_as_json->'fields', field.velocity_var_name) IS NOT NULL
```

Required Update:

The query above does not yet account for the content_identifier and the uniquePerSite field variable. We need to update this query to support these fields for the final fix.

### Acceptance Criteria

- Create Task Upgrade to create/populate unique_fields table
- Don't include the new Task Upgrade in TaskLocatorUtil, we are going to do this later

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by freddyDOTCMS**
### Proposed Changes
* Create new Task Upgrade to create and popualate the new unique_fields table

Create

https://github.com/dotCMS/core/pull/30301/files#diff-93fb205d6a0d941cfc0933bbfbfc3dd5bc70723d3fc8a8513f8058939ec7f901R113

and populate

https://github.com/dotCMS/core/pull/30301/files#diff-93fb205d6a0d941cfc0933bbfbfc3dd5bc70723d3fc8a8513f8058939ec7f901R116

it is not add in the TaskLocatorUtil, I am going to it in a future PR

This PR fixes: #30275

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    private void insertUniqueFieldsRegister(final String hash, final Map<String, Object> supportingValues) throws DotDataException {
+        new DotConnect().setSQL(INSERT_UNIQUE_FIELDS_QUERY)
+                .addParam(hash)
+                .addJSONParam(supportingValues)
+                .loadObjectResults();
+    }
+
+    private static String caculateHash(final Map<String, Object> uniqueFieldsValue) throws DotDataException {
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java:157]: Typo here

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+import static com.dotcms.util.CollectionsUtils.list;
+
+/**
+ * Task Upgrade to Create/populate the unique_fields Table for Unique Field Validation
+ *
+ * This introduces a new table, unique_fields, which will be used to validate fields that must be unique.
+ *
+ * <b>Table Structure</b>
+ *
+ * <cpde>
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java:36]: Typo here

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        createUniqueFieldTable();
+
+        try {
+            populate();
+        } catch (SQLException e) {
+            throw new DotDataException(e);
+        }
+    }
+
+    private void populate() throws DotDataException, SQLException {
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java:122]: Missing Javadoc in all methods below

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

**Review by dsilvam — CHANGES_REQUESTED** [review]
We're missing the table create script in the postgres.sql file

**Comment by freddyDOTCMS** [comment]
> We're missing the table create script in the postgres.sql file

I added that here #30306

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java (hunk 3 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java:128]: done https://github.com/dotCMS/core/pull/30301/files#diff-93fb205d6a0d941cfc0933bbfbfc3dd5bc70723d3fc8a8513f8058939ec7f901R122

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java (hunk 1 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java:176]: done https://github.com/dotCMS/core/pull/30301/files#diff-93fb205d6a0d941cfc0933bbfbfc3dd5bc70723d3fc8a8513f8058939ec7f901R176

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java (hunk 2 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task241007CreateUniqueFieldsTable.java]: done https://github.com/dotCMS/core/pull/30301/commits/1a69975f2bdf730632eed3085c3e2b4c21fca512#diff-93fb205d6a0d941cfc0933bbfbfc3dd5bc70723d3fc8a8513f8058939ec7f901R36

**Review by jcastro-dotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Review by erickgonzalez — CHANGES_REQUESTED** [review]
Shouldn't the populate be on batches? The customer who raised this issue have over 1.2M live/working contentlets and the content type which they complain have over 400K contentlets.

**Review by erickgonzalez — APPROVED** [review]

**Comment by freddyDOTCMS** [comment]
> Shouldn't the populate be on batches? The customer who raised this issue have over 1.2M live/working contentlets and the content type which they complain have over 400K contentlets.



> Shouldn't the populate be on batches? The customer who raised this issue have over 1.2M live/working contentlets and the content type which they complain have over 400K contentlets.

done https://github.com/dotCMS/core/pull/30301/commits/4ac8626cf52df39b194aec9236f9dadccc0d6a5e

**Comment by spbolton** [comment]
There is a merge issue here that is not a flakey test, but an actual merge problem.   A PR has been merged to master since this branch was updated that when merged with this code has caused it to have a reference to Task241007CreateUniqueFieldsTableTest but the import for the class is missing.   If this branch is updated the same issue will show here and will need to be resolved.  This type of error will occasionally happen and is the trade-off for not forcing the branch to be updated and retested with every merge to main.   I will update the branch now so it will force the code to be fixed before it can be merged again.

**Pull request merged**